File: stack/validParentheses_20.js
/*
LC 20 Valid Parentheses

Given a string s containing just the characters '(', ')', '{', '}', 
'[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.

Pseudocode: 

1. if you have an opening bracket, then stack it.

2. if you have a closing backet matching the top of your stack, 
    then pop your stack
3. else it means that your string is invalid

4. if you're out of the loop and you still have something on your stack, 
  it means you didn't close all your brackets which means it is invalid. 

  If it is empty, the all brackets are closed which means it is valid.



*/

/**
 * @param {string} s
 * @return {boolean}
*/

var isValid = function(s) {

  let stack = [];
  let parensMap = new Map()

  parensMap.set('(',')')
  parensMap.set('{','}')
  parensMap.set('[',']')

  
  for( let i = 0; i < s.length; i++) {

    let nextOne = stack.length - 1;
    let currentBracket = s[i];

    // if you have an opening bracket, stack it.
    if( parensMap.get(currentBracket) ) {
      stack.push( currentBracket ) 
    } 
    // if you have a matching closing bracket to what's in the stack, pop it 
    else if( parensMap.get[nextOne] === parensMap.get(currentBracket) ) {
      stack.pop()
    }
    // no matching, so invalid
    else {
      return false; 
    }

    console.log(stack)

  }

  if( stack.length === 0 ) return true;

  return false; 
    
};

console.log( isValid( "{}{" ) )

File: python_solutions/stack/valid_parentheses_20.py
# Valid Parentheses LC 20 


class Solution(object):
    def isValid(self, s):
        stack = []
        parens_dict = {
            '(': ')',
            '[': ']',
            '{': '}'
        }

        for bracket in s:
            # need to check if stack is empty or python will throw an index error 
            top_of_stack = stack[-1] if stack else None

            # if the current character is an opening bracket (because it matches
            # the KEY in the dictionary
            if bracket in parens_dict:
                # push the bracket to the stack
                stack.append(bracket)
            # else, 
            elif stack and bracket == parens_dict[top_of_stack]:
                # if it does pop the opening bracket off the stack 
                stack.pop()
            else:
                # # If it doesn't match or there's no opening bracket in the stack, return False.
                return False 

        #if the stack is empty here then return true, the parens are balanced.
        return not stack

#shorter:
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        parens_map = { "(":")", "[":"]","{":"}" }

        for bracket in s:
            if bracket in parens_map:
                stack.append(bracket)
            # you are comparing the lookup of **what would be expected closing bracket**
            # to the bracket you're currently reading in the loop
            # if they match, you have a closing bracket, else, something isn't balanced so afalse 
            elif stack and bracket == parens_map[stack[-1]]:
                stack.pop()
            # the else checks for improper sequencing or invalid characters,
            else:
                return False

        # checks for balance of parens 
        return not stack 
    
